runtime/bcutil/cfreader.c

```diff
@@ -2167,10 +2167,12 @@ checkAttributes(J9CfrClassFile* classfile, J9CfrAttribute** attributes, U_32 att
 				/* Check class name integrity? */
 			}
 
-			/* Duplicate class check */
+			/* Duplicate class check except the direct circular entries */
 			for(j = 0; j < classes->numberOfClasses; j++) {
 				for(k = j + 1; k < classes->numberOfClasses; k++) {
-					if (classes->classes[j].innerClassInfoIndex == classes->classes[k].innerClassInfoIndex) {
+					if ((classes->classes[j].innerClassInfoIndex == classes->classes[k].innerClassInfoIndex)
+					&& (classes->classes[j].innerClassInfoIndex != classes->classes[j].outerClassInfoIndex)
```

Comment:
This probably doesn't belong inside the `k` loop. Also, this code is technically incorrect (it was before your change) as it assumes that any given class will only appear at a single index, which as far as I know is not required.

Reply:
I don't know these structures well enough to be able to decide if this is correct or not. What do these indices describe?

Reply:
We are trying to find the following case in the new code here.
```
   #7 = Class              #8             // InnerClasses
   #8 = Utf8               A
  #14 = Class              #8             // InnerClasses
...
  #16 = Utf8               innerclass

InnerClasses:
  #16= #14 of #7;                         // innerclass=class A of class A
```
But the other case as follows (duplicate entries) should also be captured with the existing code:
```
   #2 = Class              #17            // A
...
   #5 = Class              #19            // B
   #6 = Utf8               innerclass

InnerClasses:
     #6= #5 of #2; //innerclass=class B of class A
     #6= #5 of #2; //innerclass=class B of class A
```

Reply:
@DanHeidinga Any comment here?

Reply:
The question is whether two different class_indexes point to two separate classes in the the constant pool which share the same UTF8 name string. If so, we might need to address that by comparing the name string.

Reply:
What does the RI do in this case?  Does it treat two distinct ClassRefs with same utf8 contents (but not index) as the same?  We should match their behaviour for this.  I'd start by validating on JDK15 and then working back to check 11 & 8 as well

Reply:
I will check with internal guys to get the RI with fix to see how it goes.